contracts/0.8.9/WithdrawalQueueBase.sol

```diff
@@ -0,0 +1,379 @@
+// SPDX-FileCopyrightText: 2023 Lido <info@lido.fi>
+// SPDX-License-Identifier: GPL-3.0
+
+/* See contracts/COMPILERS.md */
+pragma solidity 0.8.9;
+
+import "@openzeppelin/contracts-v4.4/utils/structs/EnumerableSet.sol";
+
+/**
+ * @title Internal logic to handle request queue.
+ *
+ * @author folkyatina
+ */
+abstract contract WithdrawalQueueBase {
+    using EnumerableSet for EnumerableSet.UintSet;
+
+    /// @notice precision base for share rate and discounting factor values in the contract
+    uint96 public constant E27_PRECISION_BASE = 1e27;
+    /// @notice discount factor value that means no discount applying
+    uint96 public constant NO_DISCOUNT = E27_PRECISION_BASE;
+
+    /// @notice structure representing a request for withdrawal.
+    struct WithdrawalRequest {
+        /// @notice sum of the all stETH submitted for withdrawals up to this request
+        uint128 cumulativeStETH;
+        /// @notice sum of the all shares locked for withdrawal up to this request
+        uint128 cumulativeShares;
+        /// @notice payable address of the recipient eth will be transferred to
+        address payable recipient;
+        /// @notice block.timestamp when the request was created
+        uint64 timestamp;
+        /// @notice flag if the request was claimed
+        bool claimed;
+    }
+
+    /// @notice structure representing a discount that is applied to request batch on finalization
+    struct Discount {
+        /// @notice last index the discount is applicable to. So, it is valid in (`previuosIndex`, `index`] range
+        uint256 indexInQueue;
+        /// @notice discount factor with 1e27 precision (0 - 100% discount, 1e27 - means no discount)
+        uint96 discountFactor;
+    }
+
```

Comment:
event WithdrawalBatchFinalized(
uint256 indexed from,
uint256 indexed to,
uint256 amountOfETH,
uint256 sharesBurned,
uint256 shareRate,
)

Reply:
also add timestamp